### PR TITLE
Improve error messages

### PR DESCRIPTION
**Pull request by MrKevinWeiss**
This overrides the good old classics such as `Resolving variable '${RESULT['result']}' failed` or `'SUCCESS' does not contain 'FAIL'` or who could forget the great `0 != 1` with a bit better messages that evaluate variables, print proper sentences, and explain what was expected s received.

A test to show some of these messages is also added to `dist/tools/output_to_xunit` and can be checked with `make robot-clean robot-test`.

EDIT: It turns out many other issues were discovered when working on this PR, so here it is.

### Add a test to show failure messages without requiring hardware
Adding the `03__error_msg.robot ` test.  To check one can simply use `make -C dist/tools/output_to_xunit/ robot-clean robot-test`.

Here we can see nice failure messages:

<details>

```
example.Error Msg :: Simple robot test that passes.                           
==============================================================================
Should Mock Success                                                   | PASS |
------------------------------------------------------------------------------
Should Not Mock Success                                               | FAIL |
Mock Return Success expected FAIL but received Success
------------------------------------------------------------------------------
Should Not Mock Key Error                                             | FAIL |
Mock Key Error does not have a result, return = {'foo': 'Success'}
------------------------------------------------------------------------------
Should Not Mock Type Error                                            | FAIL |
Mock Type Error does not have a result, return = Success
------------------------------------------------------------------------------
Should API Call Repeat on Timeout Success                             | PASS |
------------------------------------------------------------------------------
Should API Call Repeat on Timeout Error                               | FAIL |
Mock Return Timeout failed after 16 times with {'result': 'Timeout'}
------------------------------------------------------------------------------
example.Error Msg :: Simple robot test that passes.                   | FAIL |
6 tests, 2 passed, 4 failed

```
</details>

### Better error messages/cleanup/fixes  GPIO tests
Some non-common error messages are improved.
Utilizing the `Skip If` keyword in Robot Framework V4, this also also refactoring some redundant functions.

### Build system fixes

It turns out there are some issues with using error. The build system now uses `FEATURE_REQUIRED` and FEATURES_PROVIDED` to enforce proper wiring.

This is also useful for the CI to determine what builds actually fail and what should be failing/skipped

### Scripting fixes

A small fix in the `output_to_xunit/output_to_xunit.py` script that prevents exceptions from being thrown if a benchmark test fails and doesn't contain an expected key.

### CODE REVIEW

**Comment by MrKevinWeiss** [comment]
this PR has grown.  The build system has been broken for a while it seems.  Hopefully this will use `FEATURES_REQUIRED` so one can `make clean` without errors...

**Comment by MrKevinWeiss** [comment]
Though having build failed or skipped xunit generation would probably be better to include in the make system it is not trivial **for me** to figure out as we are using the `FEATURES_*` resolution in RIOT and I don't want to change riot.  For the moment it can simply exist in the Jenkins job, this means that it will only get generated from the CI and reproducing locally is not automatic.

Since this is a fairly simple take to do manually maybe it is fine.

I will clean up this PR and maybe we can get it in since it contains fixes and cleanups that are needed for other ongoing work.

**Comment by MrKevinWeiss** [comment]
The REMOVEME gpio test is removed so we won't see the build failed but I guess this is also out of the scope of the PR (more for the CI PR).

[Results](https://ci.riot-os.org/hil/blue/organizations/jenkins/staging_rf_tests/detail/staging_rf_tests/2/pipeline)

File: tests/Makefile.tests_common [hunk 1]
```diff
@@ -9,6 +9,23 @@ BUILD_DIR ?= $(TESTBASE)/build
...
+# list implemented in RIOT. This allows us to now use the
+# FEATURES_REQUIRED in out tests to require certain pinouts being
+# set for that board.
+define _populate_features_provided =
+  ifneq (,$(2))
+  FEATURES_PROVIDED += $(1)
+  endif
+endef
+_HIL_FEATURES_AVAILABLE = $(filter HIL_%,$(.VARIABLES))
+$(foreach v,$(_HIL_FEATURES_AVAILABLE), $(eval $(call _populate_features_provided, $(v), $($(v)))))
```
Comment (leandrolanzieri) [inline tests/Makefile.tests_common:26]: What about ?
```suggestion
_HIL_FEATURES_AVAILABLE = $(filter HIL_%,$(.VARIABLES))
FEATURES_PROVIDED += $(foreach v,$(_HIL_FEATURES_AVAILABLE),$(if $($(v)), $(v)))
```

Comment (MrKevinWeiss) [reply tests/Makefile.tests_common]: Yup

**Comment by MrKevinWeiss** [comment]
[Jenkins](https://ci.riot-os.org/hil/blue/organizations/jenkins/staging_rf_tests/detail/staging_rf_tests/3/pipeline)

**Comment by MrKevinWeiss** [comment]
[Results](https://hil.riot-os.org/results/staging_rf_tests/3/overview) are looking as expected.

**Comment by MrKevinWeiss** [comment]
@leandrolanzieri can we merge?

**Comment by MrKevinWeiss** [comment]
Hmmm something is wrong with the logs...

**Comment by MrKevinWeiss** [comment]
[This](https://ci.riot-os.org/hil/blue/organizations/jenkins/staging_rf_tests/detail/staging_rf_tests/3/pipeline/2330) Shows `spi (3.3) could not toggle to low, Check wiring` vs [this](https://ci.riot-os.org/hil/blue/organizations/jenkins/staging_rf_tests/detail/staging_rf_tests/4/pipeline/292) showing `1 != 0`.

But I still have to look into why the logs are not being generated...

**Comment by MrKevinWeiss** [comment]
Hmmm I don't think the issue of the report is with PR but we should wait until it is figured out. ping @ozfox

**Comment by MrKevinWeiss** [comment]
Figured it out, it was just not rebasing the frontend.  [Here](https://hil.riot-os.org/results/staging_rf_tests/10/overview) are the results.

We have [`PHiLIP.Write Reg expected Success but received Error`](https://hil.riot-os.org/results/staging_rf_tests/10/overview) and [`
spi (3.2) could not toggle to low, Check wiring`](https://hil.riot-os.org/results/staging_rf_tests/10/z1/tests_periph_gpio/log#s1-s2-t7-k2-k15)

**Review by leandrolanzieri — APPROVED** [review]

**Pull request merged**